Helm Release Support (#1023)
pulumi/pulumi-hugo: themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md

---
title: "Helm Release Resource for Kubernetes Generally Available"

> h1: "Announcing General Availability of Helm Release Resource for Kubernetes"

Author: Fixed.

In contrast, the new [Helm Release]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release">}}) resource uses the Helm SDK natively to perform all Helm operations. This comes with some inherrent advantages:

1. Support for the full spectrum of Helm features
2. Ability to import existing Helm releases into Pulumi state
3. Interact with the installed Helm Release through the Helm CLI

## Which Helm Resource Should I Choose?

Both the `Chart` and `Helm Release` resource are complimentary with their own pros and cons. Depending on your use case, one might be a better fit than the other. We now have a [guide to help you choose]({{< relref "/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case/">}}) the right resource for your use case.

## How Do I Get Started?

If you aren't already familiar with using Pulumi with Kubernetes, head on over to [Pulumi's getting started guide for Kubernetes]({{< relref "/docs/get-started/kubernetes/" >}}) first.

1. In your chosen Kubernetes Pulumi project, make sure the referenced version of the Pulumi-Kubernetes SDK is at least `v3.15.0`.

> 2. Helm Release support is configured to target the relevant cluster pointed to by the Kubernetes provider associated with it and no additional configuration is required by default. However, more advanced configuration options are supported, see the provider configuration [documentation]({{< relref "/registry/packages/kubernetes/api-docs/provider/#helmdriver_nodejs" >}}) for more details.

Reviewer: ```suggestion
2. Like other Kubernetes resources managed by Pulumi, the target cluster for Helm Release is configured with a [Provider]({{< relref "/registry/packages/kubernetes/api-docs/provider">}}). The default options work for typical cases, and we also provide [advanced configuration options]({{< relref "/registry/packages/kubernetes/api-docs/provider/#helmdriver_nodejs" >}}) for private registries, etc.
```

3. Refer to the instructions for your chosen language in the [resource documentation]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release/#create" >}}) to create a Helm Release resource.

> Lets look at a concrete example of Helm Release in action. In the following snippet we install [Redis](https://redis.io/) through a Helm Chart in each of the supported Pulumi languages using the new Helm Release resource:

Reviewer: ```suggestion
Let's look at a concrete example of Helm Release in action. In the following snippet we install [Redis](https://redis.io/) using the new Helm Release resource:
```

{{% choosable language typescript %}}

```typescript
import * as random from "@pulumi/random";
import * as pulumi from "@pulumi/pulumi";
import * as k8s from "@pulumi/kubernetes";

// Create a password
const pass = new random.RandomPassword("pass", {length:10});
const redisPassword = pass.result;

const namespace = new k8s.core.v1.Namespace("redis-ns");

const release = new k8s.helm.v3.Release("redis-helm", {
    chart: "redis",
    repositoryOpts: {
        repo: "https://charts.bitnami.com/bitnami",
    },
    version: "13.0.0",
    namespace: namespace.metadata.name,
    // Values from Chart's parameters specified hierarchically,
    // see https://artifacthub.io/packages/helm/bitnami/redis/13.0.0#parameters for reference.
    values: {
        cluster: {
            enabled: true,
            slaveCount: 3,
        },
        metrics: {
            enabled: true,
            service: {
                annotations: {
                    "prometheus.io/port": "9127",
                }
            },
        },
        global: {
            redis: {
                password: redisPassword,
            }
        },
        rbac: {
            create: true,
        }
    },

>     // By default Release resource will wait till all created resources

Reviewer: ```suggestion
    // By default Release resource will wait until all created resources
```

    // are available. Set this to true to skip waiting on resources being
    // available.
    skipAwait: false,
});

> // We can lookup resources once the release is installed. The release's

Reviewer: ```suggestion
// We can look up resources once the release is installed. The release's
```

		if err != nil {
			return err
		}

		redisPassword, err := random.NewRandomPassword(ctx, "pass", &random.RandomPasswordArgs{
			Length: pulumi.Int(10),
		})
		if err != nil {
			return err
		}

		rel, err := helm.NewRelease(ctx, "redis-helm", &helm.ReleaseArgs{
			Chart: pulumi.String("redis"),
			RepositoryOpts: helm.RepositoryOptsArgs{
				Repo: pulumi.String("https://charts.bitnami.com/bitnami"),
			},
			Version:   pulumi.String("13.0.0"),
			Namespace: namespace.Metadata.Name(),

            // Values from Chart's parameters specified hierarchically,
            // see https://artifacthub.io/packages/helm/bitnami/redis/13.0.0#parameters for reference.
			Values: pulumi.Map{
				"cluster": pulumi.Map{
					"enabled": pulumi.Bool(true),
				},
				"metrics": pulumi.Map{
					"enabled": pulumi.Bool(true),
					"service": pulumi.Map{
						"annotations": pulumi.StringMap{
							"prometheus.io/port": pulumi.String("9127"),
						},
						"type": pulumi.String("ClusterIP"),
					},
				},
				"global": pulumi.Map{
					"redis": pulumi.Map{
						"password": redisPassword.Result,
					},
				},
				"rbac": pulumi.BoolMap{
					"create": pulumi.Bool(true),
				},
			},

> 			// By default Release resource will wait till all created resources

Reviewer: ```suggestion
			// By default Release resource will wait until all created resources
```

			// are available. Set this to true to skip waiting on resources being
			// available.
			SkipAwait: pulumi.BoolPtr(false),
		})
		if err != nil {
			return err
		}

> 		// We can lookup resources once the release is installed. The release's

Reviewer: ```suggestion
		// We can look up resources once the release is installed. The release's
```

		// status field is set once the installation completes, so this, combined
		// with `skipAwait: pulumi.BoolPtr(false)` above, will wait to retrieve

>         //  the Redis master ClusterIP till all resources in the Chart are available.

Reviewer: ```suggestion
        //  the Redis master ClusterIP until all resources in the Chart are available.
```

```python
import pulumi
from pulumi import Output
from pulumi_random.random_password import RandomPassword
from pulumi_kubernetes.core.v1 import Namespace, Service
from pulumi_kubernetes.helm.v3 import Release, ReleaseArgs, RepositoryOptsArgs

namespace = Namespace("redis-ns")

redis_password = RandomPassword("pass", length=10)

release_args = ReleaseArgs(
    chart="redis",
    repository_opts=RepositoryOptsArgs(
        repo="https://charts.bitnami.com/bitnami"
    ),
    version="13.0.0",
    namespace=namespace.metadata["name"],

    # Values from Chart's parameters specified hierarchically,
    # see https://artifacthub.io/packages/helm/bitnami/redis/13.0.0#parameters
    # for reference.
    values={
        "cluster": {
            "enabled": True,
            "slaveCount": 3,
        },
        "metrics": {
            "enabled": True,
            "service": {
                "annotations": {
                    "prometheus.io/port": "9127",
                }
            },
        },
        "global": {
            "redis": {
                "password": redis_password.result,
            }
        },
        "rbac": {
            "create": True,
        },
    },

>     # By default Release resource will wait till all created resources

Reviewer: ```suggestion
    # By default Release resource will wait until all created resources
```

    # are available. Set this to true to skip waiting on resources being
    # available.
    skip_await=False)

release = Release("redis-helm", args=release_args)

> # We can lookup resources once the release is installed. The release's

Reviewer: ```suggestion
# We can look up resources once the release is installed. The release's
```

# status field is set once the installation completes, so this, combined
# with `skip_await=False` above, will wait to retrieve the Redis master

> # ClusterIP till all resources in the Chart are available.

Reviewer: ```suggestion
# ClusterIP until all resources in the Chart are available.
```

            // see https://artifacthub.io/packages/helm/bitnami/redis/13.0.0#parameters
            // for reference.
            Values = new InputMap<object>
            {
                ["cluster"] = new Dictionary<string,object>
                {
                    ["enabled"] = true,
                    ["slaveCount"] = 3,
                },
                ["metrics"] = new Dictionary<string,object>
                {
                    ["enabled"] = true,
                    ["service"] = new Dictionary<string, object>
                    {
                        ["annotations"] = new Dictionary<string, string>
                        {
                            ["prometheus.io/port"] = "9127",
                        },
                    }
                },
                ["global"] = new Dictionary<string,object>
                {
                    ["redis"] = new Dictionary<string, object>
                    {
                        ["password"] = redisPassword.Result,
                    },
                },
                ["rbac"] = new Dictionary<string,object>
                {
                    ["create"] = true,
                }
            },

>             // By default Release resource will wait till all created resources

Reviewer: ```suggestion
            // By default Release resource will wait until all created resources
```

            // are available. Set this to true to skip waiting on resources being
            // available.
            SkipAwait = false,
        };

        var release = new Release("redis-helm", releaseArgs);

>         // We can lookup resources once the release is installed. The release's

Reviewer: ```suggestion
        // We can look up resources once the release is installed. The release's
```

        // status field is set once the installation completes, so this, combined
        // with `skip_await=False` above, will wait to retrieve the Redis master

>         // ClusterIP till all resources in the Chart are available.

Reviewer: ```suggestion
        // ClusterIP until all resources in the Chart are available.
```
